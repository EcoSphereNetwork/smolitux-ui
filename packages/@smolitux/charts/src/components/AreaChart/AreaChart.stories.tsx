import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { AreaChart } from './AreaChart';

const meta: Meta<typeof AreaChart> = {
  title: 'Charts/AreaChart',
  component: AreaChart,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    height: { control: { type: 'number' } },
    width: { control: { type: 'text' } },
    showGrid: { control: { type: 'boolean' } },
    showPoints: { control: { type: 'boolean' } },
    showTooltips: { control: { type: 'boolean' } },
    showLegend: { control: { type: 'boolean' } },
    animated: { control: { type: 'boolean' } },
    startYAxisAtZero: { control: { type: 'boolean' } },
    smooth: { control: { type: 'boolean' } },
    filled: { control: { type: 'boolean' } },
    stacked: { control: { type: 'boolean' } },
  },
};

export default meta;
type Story = StoryObj<typeof AreaChart>;

// Beispieldaten für eine einzelne Serie
const singleSeriesData = {
  id: 'pageViews',
  name: 'Seitenaufrufe',
  data: [
    { x: 'Jan', y: 100 },
    { x: 'Feb', y: 150 },
    { x: 'Mär', y: 200 },
    { x: 'Apr', y: 120 },
    { x: 'Mai', y: 180 },
    { x: 'Jun', y: 250 },
  ]
};

// Beispieldaten für mehrere Serien
const multiSeriesData = [
  {
    id: 'pageViews',
    name: 'Seitenaufrufe',
    data: [
      { x: 'Jan', y: 100 },
      { x: 'Feb', y: 150 },
      { x: 'Mär', y: 200 },
      { x: 'Apr', y: 120 },
      { x: 'Mai', y: 180 },
      { x: 'Jun', y: 250 },
    ]
  },
  {
    id: 'uniqueVisitors',
    name: 'Eindeutige Besucher',
    data: [
      { x: 'Jan', y: 50 },
      { x: 'Feb', y: 90 },
      { x: 'Mär', y: 120 },
      { x: 'Apr', y: 80 },
      { x: 'Mai', y: 110 },
      { x: 'Jun', y: 150 },
    ]
  }
];

export const Default: Story = {
  args: {
    data: singleSeriesData,
    height: 300,
    width: 600,
    title: 'Seitenaufrufe pro Monat',
    showGrid: true,
    showPoints: true,
    showTooltips: true,
    showLegend: true,
    animated: true,
    startYAxisAtZero: true,
    smooth: false,
    filled: true,
  },
};

export const MultiSeries: Story = {
  args: {
    data: multiSeriesData,
    height: 300,
    width: 600,
    title: 'Website-Statistiken',
    showGrid: true,
    showPoints: true,
    showTooltips: true,
    showLegend: true,
    legendPosition: 'bottom',
    animated: true,
    startYAxisAtZero: true,
    smooth: false,
    filled: true,
  },
};

export const Stacked: Story = {
  args: {
    data: multiSeriesData,
    height: 300,
    width: 600,
    title: 'Gestapelte Website-Statistiken',
    showGrid: true,
    showPoints: false,
    showTooltips: true,
    showLegend: true,
    legendPosition: 'right',
    animated: true,
    startYAxisAtZero: true,
    smooth: false,
    filled: true,
    stacked: true,
  },
};

export const Smooth: Story = {
  args: {
    data: singleSeriesData,
    height: 300,
    width: 600,
    title: 'Geglättete Kurve',
    showGrid: true,
    showPoints: true,
    showTooltips: true,
    showLegend: false,
    animated: true,
    startYAxisAtZero: true,
    smooth: true,
    filled: true,
  },
};

export const CustomColors: Story = {
  args: {
    data: multiSeriesData,
    height: 300,
    width: 600,
    title: 'Benutzerdefinierte Farben',
    showGrid: true,
    showPoints: true,
    showTooltips: true,
    showLegend: true,
    animated: true,
    startYAxisAtZero: true,
    smooth: false,
    filled: true,
    colors: ['#FF6B6B', '#4ECDC4'],
  },
};

export const WithAxisLabels: Story = {
  args: {
    data: singleSeriesData,
    height: 300,
    width: 600,
    title: 'Mit Achsenbeschriftungen',
    axisLabels: { x: 'Monat', y: 'Anzahl' },
    units: { y: 'Besucher' },
    showGrid: true,
    showPoints: true,
    showTooltips: true,
    showLegend: true,
    animated: true,
    startYAxisAtZero: true,
    smooth: false,
    filled: true,
  },
};