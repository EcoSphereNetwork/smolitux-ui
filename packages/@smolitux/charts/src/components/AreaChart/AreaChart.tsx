// packages/@smolitux/charts/src/components/AreaChart/AreaChart.tsx
import React, { useMemo, forwardRef } from 'react';
import { useTheme } from '@smolitux/theme';

export interface AreaChartDataPoint {
  /** X-Wert (Label) */
  x: string | number;
  /** Y-Wert (Datenpunkt) */
  y: number;
  /** Optional: Kategorie für Multi-Serien Charts */
  category?: string;
  /** Optional: Metadata für Tooltips etc. */
  metadata?: Record<string, any>;
}

export interface AreaChartSeries {
  /** ID der Serie */
  id: string;
  /** Name der Serie (für Legende) */
  name: string;
  /** Farbe der Serie */
  color?: string;
  /** Füllung (z.B. für Gradient) */
  fill?: string;
  /** Daten der Serie */
  data: AreaChartDataPoint[];
  /** Liniendicke */
  strokeWidth?: number;
  /** Durchschnittslinie anzeigen */
  showAverage?: boolean;
  /** Gestapelt (für mehrere Serien) */
  stacked?: boolean;
}

export interface AreaChartProps extends Omit<React.SVGProps<SVGSVGElement>, 'data'> {
  /** Einzelne Datenserie oder Array von Serien */
  data: AreaChartSeries | AreaChartSeries[];
  /** Höhe des Charts */
  height?: number;
  /** Breite des Charts (Default: 100%) */
  width?: number | string;
  /** Padding innerhalb des Charts */
  padding?: { top?: number; right?: number; bottom?: number; left?: number };
  /** Chart-Titel */
  title?: string;
  /** Achsentitel */
  axisLabels?: { x?: string; y?: string };
  /** Einheiten für Achsenbeschriftungen */
  units?: { x?: string; y?: string };
  /** Grid-Linien anzeigen */
  showGrid?: boolean;
  /** Datenpunkte anzeigen */
  showPoints?: boolean;
  /** Tooltips anzeigen */
  showTooltips?: boolean;
  /** Legende anzeigen */
  showLegend?: boolean;
  /** Position der Legende */
  legendPosition?: 'top' | 'right' | 'bottom' | 'left';
  /** Animation aktivieren */
  animated?: boolean;
  /** Y-Achse bei Null beginnen */
  startYAxisAtZero?: boolean;
  /** Angepasste Farben für mehrere Serien */
  colors?: string[];
  /** Angepasste Formatierung für Y-Achsenbeschriftungen */
  formatYLabel?: (value: number) => string;
  /** Angepasste Formatierung für X-Achsenbeschriftungen */
  formatXLabel?: (value: string | number) => string;
  /** Für responsive SVG (viewBox) */
  aspectRatio?: number;
  /** Kurven glätten */
  smooth?: boolean;
  /** Füllung für Bereich unter der Kurve */
  filled?: boolean;
  /** Gestapelte Bereiche (für mehrere Serien) */
  stacked?: boolean;
  /** Benutzerdefiniertee Tooltip-Rendering */
  renderTooltip?: (point: AreaChartDataPoint, series: AreaChartSeries) => React.ReactNode;
  /** Unteres Limit für den Bereich (für Bereichsdiagramme mit oberem und unterem Limit) */
  areaBaseValue?: number;
}

/**
 * AreaChart-Komponente für Zeitreihen und kontinuierliche Daten
 * 
 * @example
 * ```tsx
 * <AreaChart
 *   data={{
 *     id: 'views',
 *     name: 'Page Views',
 *     data: [
 *       { x: 'Jan', y: 100 },
 *       { x: 'Feb', y: 150 },
 *       { x: 'Mar', y: 200 },
 *       { x: 'Apr', y: 120 },
 *       { x: 'May', y: 180 },
 *     ]
 *   }}
 *   height={300}
 *   showGrid
 *   showPoints
 *   filled
 * />
 * ```
 */
export const AreaChart = forwardRef<SVGSVGElement, AreaChartProps>(({
  data,
  height = 300,
  width = '100%',
  padding = { top: 30, right: 30, bottom: 40, left: 50 },
  title,
  axisLabels,
  units,
  showGrid = true,
  showPoints = true,
  showTooltips = true,
  showLegend = true,
  legendPosition = 'top',
  animated = true,
  startYAxisAtZero = true,
  colors,
  formatYLabel = (value) => `${value}`,
  formatXLabel = (value) => `${value}`,
  aspectRatio = 16 / 9,
  smooth = false,
  filled = true,
  stacked = false,
  renderTooltip,
  areaBaseValue,
  className = '',
  ...rest
}, ref) => {
  const theme = useTheme() as any;
  const themeMode = theme.themeMode ?? 'light';

  
  // Sicherstellen, dass wir mit einem Array von Serien arbeiten
  const seriesArray = Array.isArray(data) ? data : [data];
  
  // Standard-Farbpalette
  const defaultColors = [
    '#3B82F6', // blue-500
    '#10B981', // emerald-500
    '#F59E0B', // amber-500
    '#EF4444', // red-500
    '#8B5CF6', // violet-500
    '#EC4899', // pink-500
    '#06B6D4', // cyan-500
    '#14B8A6', // teal-500
    '#F97316', // orange-500
    '#6366F1', // indigo-500
  ];
  
  // Tatsächlich verwendete Farben
  const chartColors = colors || defaultColors;
  
  // Padding-Werte mit Defaults
  const paddingTop = padding.top ?? 30;
  const paddingRight = padding.right ?? 30;
  const paddingBottom = padding.bottom ?? 40;
  const paddingLeft = padding.left ?? 50;
  
  // Eigentliche Zeichenfläche (ohne Padding)
  const drawingWidth = typeof width === 'number' ? width - paddingLeft - paddingRight : '100%';
  const drawingHeight = height - paddingTop - paddingBottom;
  
  // SVG viewBox für Responsivität
  const viewBoxWidth = 1000;
  const viewBoxHeight = Math.round(viewBoxWidth / aspectRatio);
  
  // Basiswert für den Bereich (default: 0)
  const baseValue = areaBaseValue !== undefined ? areaBaseValue : (startYAxisAtZero ? 0 : undefined);
  
  // Datenaufbereitung und Skalierung
  const {
    xLabels,
    yMin,
    yMax,
    normalizedSeries,
    yScale,
  } = useMemo(() => {
    // Alle X-Labels aus allen Serien extrahieren (für Achsenbeschriftung)
    const allXValues = seriesArray
      .flatMap(series => series.data.map(d => d.x))
      .filter((value, index, self) => self.indexOf(value) === index)
      .sort((a, b) => {
        if (typeof a === 'number' && typeof b === 'number') {
          return a - b;
        }
        return String(a).localeCompare(String(b));
      });
    
    // Min/Max Y-Werte für Skalierung
    let minY = startYAxisAtZero ? 0 : Number.MAX_VALUE;
    let maxY = Number.MIN_VALUE;
    
    if (stacked) {
      // Bei gestapelten Bereichen müssen wir die Summen pro X-Wert berechnen
      const stackTotals = new Map<string | number, number>();
      
      allXValues.forEach(x => {
        let sum = 0;
        seriesArray.forEach(series => {
          const point = series.data.find(d => d.x === x);
          if (point) sum += point.y;
        });
        stackTotals.set(x, sum);
        if (sum > maxY) maxY = sum;
      });
      
      if (!startYAxisAtZero) {
        minY = Math.min(...stackTotals.values());
      }
    } else {
      // Normaler Min/Max über alle Serien
      seriesArray.forEach(series => {
        series.data.forEach(point => {
          if (point.y < minY) minY = point.y;
          if (point.y > maxY) maxY = point.y;
        });
      });
    }
    
    // Wenn der Basis-Wert für den Bereich definiert ist, berücksichtigen wir ihn
    if (baseValue !== undefined && baseValue < minY) {
      minY = baseValue;
    }
    
    // Sicherheitscheck für den Fall, dass alle Werte gleich sind
    if (minY === maxY) {
      if (minY === 0) {
        maxY = 1; // Wenn alle Werte 0 sind
      } else {
        // Andernfalls etwas Spielraum schaffen
        minY = startYAxisAtZero ? 0 : minY * 0.9;
        maxY = maxY * 1.1;
      }
    }
    
    // Y-Skala für die Normalisierung der Werte
    const yScale = (value: number) => {
      return 1 - ((value - minY) / (maxY - minY));
    };
    
    // Normalisierte Daten für die Darstellung
    const normalizedSeries = seriesArray.map((series, index) => {
      return {
        ...series,
        color: series.color || chartColors[index % chartColors.length],
        data: series.data.map(point => ({
          ...point,
          normalizedY: yScale(point.y)
        }))
      };
    });
    
    return {
      xLabels: allXValues,
      yMin: minY,
      yMax: maxY,
      normalizedSeries,
      yScale
    };
  }, [seriesArray, startYAxisAtZero, stacked, baseValue, chartColors]);
  
  // Hier würde die eigentliche Rendering-Logik folgen
  // Für dieses Beispiel geben wir nur ein Platzhalter-SVG zurück
  return (
    <svg
      ref={ref}
      width={width}
      height={height}
      viewBox={`0 0 ${viewBoxWidth} ${viewBoxHeight}`}
      className={`smolitux-area-chart ${className}`}
      {...rest}
    >
      <rect
        x="0"
        y="0"
        width={viewBoxWidth}
        height={viewBoxHeight}
        fill="none"
        stroke={themeMode === 'dark' ? '#374151' : '#E5E7EB'}
        strokeWidth="1"
      />
      <text
        x={viewBoxWidth / 2}
        y={viewBoxHeight / 2}
        textAnchor="middle"
        dominantBaseline="middle"
        fill={themeMode === 'dark' ? '#D1D5DB' : '#4B5563'}
        fontSize="24"
      >
        AreaChart Platzhalter
      </text>
    </svg>
  );
});

export default AreaChart;