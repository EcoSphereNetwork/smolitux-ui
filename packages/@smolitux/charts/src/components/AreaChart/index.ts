// packages/@smolitux/charts/src/components/AreaChart/index.ts
export { default, type AreaChartProps, type AreaChartDataPoint, type AreaChartSeries } from './AreaChart';
