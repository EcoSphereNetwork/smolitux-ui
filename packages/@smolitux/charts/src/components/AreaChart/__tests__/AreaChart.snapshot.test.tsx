import React from 'react';
import { render } from '@testing-library/react';
import { AreaChart } from '../AreaChart';

jest.mock('@smolitux/theme', () => ({
  useTheme: () => ({ themeMode: 'light' }),
}));

describe('AreaChart Snapshots', () => {
  const mockData = {
    id: 'testSeries',
    name: 'Test Data',
    data: [
      { x: 'Jan', y: 100 },
      { x: 'Feb', y: 150 },
      { x: 'Mar', y: 200 },
      { x: 'Apr', y: 120 },
      { x: 'May', y: 180 },
    ],
  };

  it('renders default chart', () => {
    const { asFragment } = render(<AreaChart data={mockData} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
