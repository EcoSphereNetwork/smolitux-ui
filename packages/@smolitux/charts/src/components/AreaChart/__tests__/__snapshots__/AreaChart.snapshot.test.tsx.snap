// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreaChart Snapshots renders default chart 1`] = `
<DocumentFragment>
  <svg
    class="smolitux-area-chart smolitux-area-chart"
    height="300"
    role="img"
    tabindex="0"
    viewBox="0 0 1000 563"
    width="100%"
  >
    <rect
      fill="#F9FAFB"
      height="230"
      stroke="#E5E7EB"
      stroke-width="1"
      width="0"
      x="50"
      y="30"
    />
    <g
      class="grid-lines"
    >
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="260"
        y2="260"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="202.5"
        y2="202.5"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="145"
        y2="145"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="87.5"
        y2="87.5"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="30"
        y2="30"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="30"
        y2="260"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="30"
        y2="260"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="30"
        y2="260"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="30"
        y2="260"
      />
      <line
        stroke="#E5E7EB"
        stroke-dasharray="4,4"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="30"
        y2="260"
      />
    </g>
    <g
      class="data-areas"
    >
      <g
        class="series-area"
      >
        <path
          d="M 50 122 L 50 87.5 L 50 145 L 50 30 L 50 53 L 50 260 L 50 260 Z"
          fill="#3B82F6"
          opacity="0.4"
          stroke="#3B82F6"
          stroke-width="2"
        />
        <circle
          cx="50"
          cy="122"
          fill="#FFFFFF"
          r="4"
          stroke="#3B82F6"
          stroke-width="2"
        />
        <circle
          cx="50"
          cy="87.5"
          fill="#FFFFFF"
          r="4"
          stroke="#3B82F6"
          stroke-width="2"
        />
        <circle
          cx="50"
          cy="145"
          fill="#FFFFFF"
          r="4"
          stroke="#3B82F6"
          stroke-width="2"
        />
        <circle
          cx="50"
          cy="30"
          fill="#FFFFFF"
          r="4"
          stroke="#3B82F6"
          stroke-width="2"
        />
        <circle
          cx="50"
          cy="53"
          fill="#FFFFFF"
          r="4"
          stroke="#3B82F6"
          stroke-width="2"
        />
      </g>
    </g>
    <g
      class="x-axis"
    >
      <line
        stroke="#9CA3AF"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="260"
        y2="260"
      />
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="50"
          x2="50"
          y1="260"
          y2="265"
        />
        <text
          fill="#4B5563"
          font-size="12"
          text-anchor="middle"
          x="50"
          y="280"
        >
          Apr
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="50"
          x2="50"
          y1="260"
          y2="265"
        />
        <text
          fill="#4B5563"
          font-size="12"
          text-anchor="middle"
          x="50"
          y="280"
        >
          Feb
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="50"
          x2="50"
          y1="260"
          y2="265"
        />
        <text
          fill="#4B5563"
          font-size="12"
          text-anchor="middle"
          x="50"
          y="280"
        >
          Jan
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="50"
          x2="50"
          y1="260"
          y2="265"
        />
        <text
          fill="#4B5563"
          font-size="12"
          text-anchor="middle"
          x="50"
          y="280"
        >
          Mar
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="50"
          x2="50"
          y1="260"
          y2="265"
        />
        <text
          fill="#4B5563"
          font-size="12"
          text-anchor="middle"
          x="50"
          y="280"
        >
          May
        </text>
      </g>
    </g>
    <g
      class="y-axis"
    >
      <line
        stroke="#9CA3AF"
        stroke-width="1"
        x1="50"
        x2="50"
        y1="30"
        y2="260"
      />
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="260"
          y2="260"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="260"
        >
          0
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="202.5"
          y2="202.5"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="202.5"
        >
          50
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="145"
          y2="145"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="145"
        >
          100
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="87.5"
          y2="87.5"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="87.5"
        >
          150
        </text>
      </g>
      <g>
        <line
          stroke="#9CA3AF"
          stroke-width="1"
          x1="45"
          x2="50"
          y1="30"
          y2="30"
        />
        <text
          dominant-baseline="middle"
          fill="#4B5563"
          font-size="12"
          text-anchor="end"
          x="40"
          y="30"
        >
          200
        </text>
      </g>
    </g>
    <g
      class="chart-legend"
    >
      <g
        transform="translate(60, 10)"
      >
        <g
          transform="translate(0, 0)"
        >
          <rect
            fill="#3B82F6"
            height="12"
            rx="2"
            width="12"
          />
          <text
            dominant-baseline="middle"
            fill="#4B5563"
            font-size="12"
            x="20"
            y="10"
          >
            Test Data
          </text>
        </g>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;
